File: round1/19.cpp
https://leetcode-cn.com/problems/remove-nth-node-from-end-of-list/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode * preHead = new ListNode(0, head);
        ListNode* queueHead = preHead, * queueTail = preHead;
        for(int i = 0; i < n; i++){
            queueTail = queueTail->next;
        } 
        while(queueTail->next != nullptr){
            queueHead = queueHead->next;
            queueTail = queueTail->next;
        }
        ListNode* temp = queueHead->next;
        queueHead->next = temp->next;
        return preHead->next;
    }
};

File: aim2OfferRound1/19.go
package main

// 快慢指针，先让快指针跑n次，然后快慢指针一起跑
// 当快指针到达链表尾，删除慢指针下一个节点即可
// 注意两个指针初始化应该是虚拟头节点，避免只有一个头节点的特殊情况

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummyHead := &ListNode{-1,head}
	slow, fast := dummyHead, dummyHead
	for n > 0 {
		n--
		fast = fast.Next
	}
	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}
	slow.Next = slow.Next.Next
	return dummyHead.Next
}
